test/utils/ktesting/tcontext.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ktesting
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"testing"
+
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/client-go/dynamic"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	_ "k8s.io/component-base/logs/testinit"
+	"k8s.io/klog/v2"
+	"k8s.io/klog/v2/ktesting"
+	"k8s.io/kubernetes/test/utils/format"
+)
+
+// TContext combines [context.Context], [testing.TB] and some additional
+// methods.  Log output is associated with the current test. Errors ([Error],
+// [Errorf]) are recorded with "ERROR" as prefix, fatal errors ([Fatal],
+// [Fatalf]) with "FATAL ERROR".
+type TContext interface {
+	context.Context
+	testing.TB
+
+	// Cancel can be invoked to cancel the context before the test is completed.
+	// Tests which use the context to control goroutines and then wait for
+	// termination of those goroutines must call Cancel to avoid a deadlock.
+	Cancel()
+
+	// Logger returns a logger for the current test.
+	Logger() klog.Logger
+
+	// GetTB returns the underlying testing.TB. This can be used to "break
+	// the glass" and cast back into a testing.T or testing.TB.
+	GetTB() testing.TB
+
+	Client() clientset.Interface
+	Dynamic() dynamic.Interface
+	APIExtensions() apiextensions.Interface
+}
```

Comment:
Wild (?) idea: what if we add enough functionality into this interface such that test/e2e/framework/* sub-packages can be called with a `tCtx` instead of a `*framework.Framework`?

Then we could have one implementation of the interface for integration tests (the one from this PR) and another for framework-based E2E tests.

That should allow us to write E2E and integration tests with the same helper packages. I bet we could remove quite a bit of duplicated code.

Perhaps it'll even be possible to implement some of Ginkgo's features (CTRL-C handling, progress reporting) also for integration tests.

The interface is useful already without those next steps. I just wanted to mention them lest I forget and to give you the whole picture - well, as much of it that I have right now :grin:

I'll have to experiment a bit more with this...

Reply:
I just found this, from 8 years ago:

https://github.com/kubernetes/kubernetes/blob/f55d18a1cf6cd9b49740819cc08479bd0d362c7b/test/e2e/framework/framework.go#L90-L91

So it's not just me who thinks that integration and E2E testing are similar...

Reply:
It really seems doable. I am including some first steps towards that as a WIP commit at the end of this PR. If we decide to do this, then it should become part of a separate PR.